### ISSUE TITLE
[Feature request] add improved resampling methods (with gamma correction)

### ISSUE THREAD

**Issue by Filoppi** [rt64/rt64#37]
With probably many gamers interested in playing N64 games at their native resolution but on modern high resolution displays, an advanced resampling method is required. Preferably, it should also be gamma corrected (linearize from gamma space, resample in linear space and re-covert to gamma space) to avoid hue shifts.

See here: [https://github.com/dolphin-emu/dolphin/pull/11999](https://github.com/dolphin-emu/dolphin/pull/11999)
and here: [https://dolphin-emu.org/blog/2023/11/25/dolphin-progress-report-august-september-and-october-2023/#50-19939-implement-output-resampling-upscalingdownscaling-by-filoppi-and-samb](https://dolphin-emu.org/blog/2023/11/25/dolphin-progress-report-august-september-and-october-2023/#50-19939-implement-output-resampling-upscalingdownscaling-by-filoppi-and-samb)

This video showcases it in motion and explains it in simpler terms: [https://www.youtube.com/watch?v=ANAYINl_6bg](https://www.youtube.com/watch?v=ANAYINl_6bg)

**Comment by DarioSamo** [comment]
> With probably many gamers interested in playing N64 games at their native resolution but on modern high resolution displays, an advanced resampling method is required.

I don't really follow your suggestion, there's already a [pixel scaling method](https://github.com/rt64/rt64/blob/main/src/shaders/VideoInterfacePS.hlsl) implemented. What options would you prefer instead? I'm not gonna add every option linked in that issue.

**Comment by Filoppi** [comment]
Of course :). I suggest simply adding "Area Resampling", which is probably the best one we have for upscaling and downscaling:
https://github.com/dolphin-emu/dolphin/blob/8167a046ef763d25e6bce91ef07abb99659dc39f/Data/Sys/Shaders/default_pre_post_process.glsl#L198
I would also suggest to do it gamma corrected, linearizing with a gamma pow of 2.35.

**Comment by DarioSamo** [comment]
> Of course :). I suggest simply adding "Area Resampling", which is probably the best one we have for upscaling and downscaling: https://github.com/dolphin-emu/dolphin/blob/8167a046ef763d25e6bce91ef07abb99659dc39f/Data/Sys/Shaders/

I'm afraid I can't simply copy implementations of GPL code as the license is incompatible with this repository (MIT). You'll need to provide a reimplementation based on the article or show that it's from a source that doesn't have that license.

Adding more options should be as simple as modifying the shader I posted with another compilation option and swapping out the VI Renderer pipeline.

**Comment by Filoppi** [comment]
That code was written from scratch for the PR I linked above, by @Sam-Belliveau and me. Let's see if we can find a way you could safely re-use it. You have my approval anyway, or I could make the PR myself if it helps.

**Comment by DarioSamo** [comment]
> That code was written from scratch for the PR I linked above, by @Sam-Belliveau and me. Let's see if we can find a way you could safely re-use it. You have my approval anyway, or I could make the PR myself if it helps.

Sure, that sounds alright to me. It should be pretty straightforward to add the new options. Whether Recomp exposes that will be a different story due to UX reasons, but I don't mind having more options in RT64 itself.

**Comment by Sam-Belliveau** [comment]
I am the original author of Area Sampling, and I have an MIT implementation, but its a little harder to read.

https://github.com/Sam-Belliveau/MKS66-Graphics-Library/blob/4ccf04f977a15007e32bdb5a238704eaaff0c895/src/graphics/Image.hpp#L126-L164

This is the code to take the weighted area sample of any section of an image. Area sampling just does this for each pixel.

Also remember, unless you use mipmapping, theres no way to properly downsample without taking multiple samples.
